### PR TITLE
Use pastebin.m.o with FormData. Fixes #86.

### LINKED ISSUE #86: copy about:support to pastebin menu entry is broken
When you try the above mentioned command you will get the following entry in the Error Console:

Error: [Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIWebNavigation.loadURI]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: chrome://global/content/bindings/browser.xml :: loadURIWithFlags :: line 156"  data: no]
Source File: chrome://browser/content/tabbrowser.xml
Line: 1358

### PR DESCRIPTION
**Pull request by xabolcs**
This is a pull request to deal with the broken feature - send `about:support` to Pastebin. (#86)

I made other improvements too: 
- disable menuitem while sending to avoid multiple pastes
- hide menuitem if there isn't any `about:support` page
- Thunderbird support - a little ugly modification was enough

### CODE REVIEW

**Comment by xabolcs** [comment]
For the record.

__whimboo__ _[commented][1] a day ago_
> [...]. When we do that lets add the new option at the place where the pastebin one is located at right now. The pastbin entry should then be separated with a line.

[1]: /mozilla/nightlytt/issues/188#issuecomment-148762258

File: extension/chrome/content/nightly.js [hunk 1]
```diff
@@ -226,20 +226,52 @@ copyTemplate: function(template) {
...
+        onLoadCallback();
+      }
     }
   };
-  request.send(postdata);
+
+  if (typeof FormData === "function") {
+    postdata = new FormData();
+    postdata.append("code2", content);
+    postdata.append("expiry", "d");
```
Comment (whimboo) [inline extension/chrome/content/nightly.js]: Hm, did I miss to give feedback here? If yes, then here it comes... I don't think we should keep forever. Maybe we use a month for now given that we do not have an option panel for NTT yet. Once we have that we should use the default of the API and allow the user to change it in the options.

**Comment by whimboo** [comment]
@xabolcs, quick question. Is this PR actually blocked by the multi-form data PR? If yes, lets indicate it by a label. Otherwise maybe we can get this done next?

**Comment by xabolcs** [comment]
> Is this PR actually blocked by the multi-form data PR?

Sorry, which one?

**Comment by whimboo** [comment]
Issue #176 which was referenced "lately" on this issue.

**Comment by xabolcs** [comment]
Ahh. That one. Thanks!

It's not blocked by #176. But I will remove the `MultipartFormData` fallback from this PR.

**Comment by whimboo** [comment]
Nice. I will also make sure to review PRs promptly within a 24h time frame from now on!

**Comment by xabolcs** [comment]
Please note that I'll update this PR tonight, so you are good enough to review it tomorrow! :+1:

**Comment by xabolcs** [comment]
> Please note that I'll update this PR tonight, ...

Sorry, didn't get enough time slice for this.

File: extension/chrome/content/nightly.js [hunk 2]
```diff
@@ -220,20 +220,71 @@ copyTemplate: function(template) {
...
+  function onError() {
+    if (typeof onErrorCallback === "function") {
+      onErrorCallback();
+    }
+  }
+
+  request.onabort = onError;
+  request.onerror = onError;
+
+  if (typeof FormData === "function") {
```
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Let die with an Exception if `FormData` isn't available.

File: extension/chrome/content/nightly.js [hunk 3]
```diff
@@ -226,20 +226,71 @@ copyTemplate: function(template) {
   nightly.copyText(nightly.generateText(nightly.getTemplate(template)));
 },
 
-pastebin: function (content) {
-  var postdata = "paste_code=" + encodeURIComponent(content);
+pastebin: function (content, onLoadCallback, onErrorCallback) {
+  var postdata;
+  const pastebinURL = "http://pastebin.mozilla.org/";
```
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Not moving to top of `nightly.js`, because it have to be referenced as `nightly.PASTEBIN_URL` or `this.PASTEBIN_URL` which doesn't have any benefits, IMHO.

Comment (xabolcs) [inline extension/chrome/content/nightly.js]: One more note here ... no need to `const` the URL anyway.

URL is provider specific, as the all other attribute like `expiry`, `content2`, `format`, etc.

**Comment by xabolcs** [comment]
@whimboo, I'm done with the refresh, but it doesn't work. :cry:

**Comment by xabolcs** [comment]
Hmmm. The second test was good. What happened?!

**Comment by xabolcs** [comment]
OK. I got it!

The very first (see Private Browsing) request ends up an empty `pastebin.mozilla.org` site.
All subsequent request will successfully deliver the `about:support`'s content.

**Comment by xabolcs** [comment]
But it works on the first request if I use `https` protocol. :+1:

**Comment by xabolcs** [comment]
@whimboo, I'm done with refreshing this PR for now.
Please review!

File: extension/chrome/content/nightly.js (hunk 3 above)
Comment (whimboo) [inline extension/chrome/content/nightly.js]: That's true. Each provider would have other values.

**Comment by whimboo** [comment]
> But it works on the first request if I use https protocol.

I see. In general everyone should use HTTPS if available. So this is a good move.

File: extension/chrome/content/nightly.js [hunk 4]
```diff
@@ -220,19 +220,43 @@ copyTemplate: function(template) {
   nightly.copyText(nightly.generateText(nightly.getTemplate(template)));
 },
 
-pastebin: function (content) {
-  var postdata = "paste_code=" + encodeURIComponent(content);
+pastebin: function (content, onLoadCallback, onErrorCallback) {
+  if (content === null || content === "" || content === undefined) {
+    onError();
```
Comment (whimboo) [inline extension/chrome/content/nightly.js]: We will loose why we call onError() here. I will be better to provide a reason why we failed.

File: extension/chrome/content/nightly.js [hunk 5]
```diff
@@ -254,16 +278,25 @@ parseHTML: function(url, callback) {
...
+
   nightly.parseHTML("about:support", function(doc) {
     var contents = doc.getElementById("contents");
     var text = nightlyPPrint.createTextForElement(contents);
-    nightly.pastebin(text);
+    function enableMenuItem() {
+      node.removeAttribute("loading");
+      node.disabled = false;
+    }
+    nightly.pastebin(text, enableMenuItem, enableMenuItem);
```
Comment (whimboo) [inline extension/chrome/content/nightly.js]: I don't think that we want to enable the menuitem when there was an error.

File: extension/chrome/content/nightly.js (hunk 4 above)
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Passing a String message "Not posting empty content!" to `onError()` will do the job?

File: extension/chrome/content/nightly.js (hunk 5 above)
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Menuitem should be enabled in all case to allow other uploads happen.
What if a temporary network outage triggers the error? Should the user restart the application to push other about:support?

Of course a feedback to User is welcome here! A popup, or at least a Console log.

File: extension/chrome/content/nightly.js (hunk 4 above)
Comment (whimboo) [inline extension/chrome/content/nightly.js]: Would the API call fail in case when the above condition is true? If that is the case I would not do anything here to prevent submitting an empty string.

File: extension/chrome/content/nightly.js (hunk 5 above)
Comment (whimboo) [inline extension/chrome/content/nightly.js]: Oh, wait. We disable it while submitting data, and enable it afterward. So this is fine.

File: extension/chrome/content/nightly.js [hunk 6]
```diff
@@ -220,19 +220,43 @@ copyTemplate: function(template) {
...
+      }
+      if (typeof onLoadCallback === "function") {
+        onLoadCallback();
+      }
     }
   };
+
+  function onError() {
+    if (typeof onErrorCallback === "function") {
+      onErrorCallback();
```
Comment (whimboo) [inline extension/chrome/content/nightly.js]: Lets make sure we pass a message through eg. the one which the API call returns in case of a failure. It can then be logged to the console.

File: extension/chrome/content/nightly.js (hunk 4 above)
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: For an empty string (`""`), the start page loads.
For `null` value, the [landing page][1] loads with `null` string inside. `null` gets converted to string.
Same for `undefinded`: it is converted to [text][2].

[1]: https://pastebin.mozilla.org/8905955
[2]: https://pastebin.mozilla.org/8905957

File: extension/chrome/content/nightly.js (hunk 6 above)
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: I left the `onErrorCallback()` and `onLoadCallback()` callbacks parameterless.

The local `onError()`'s `message` parameter is processed, and showed as a popup if needed.

File: extension/chrome/content/nightly.js (hunk 4 above)
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Should I allow `null` `undefined`?

Comment (whimboo) [inline extension/chrome/content/nightly.js]: I would leave it all up to the RestAPI to decide what's valid or not. It knows it best. We should just make sure to handle possible failures as returned correctly.

File: extension/chrome/content/nightly.js (hunk 6 above)
Comment (whimboo) [inline extension/chrome/content/nightly.js]: Where is this done? I cannot find this code.

Comment (whimboo) [inline extension/chrome/content/nightly.js]: Strange. I do not see the updated code from the last commit in this PR!

File: extension/chrome/content/nightly.js [hunk 7]
```diff
@@ -220,19 +220,50 @@ copyTemplate: function(template) {
...
+
+  function onError(message) {
+    if (typeof onErrorCallback === "function") {
+      onErrorCallback();
+    }
+    if (message === null || message === "" || message === undefined) {
+      return;
+    }
+    if (typeof message !== "string") {
+      message = "An error occured while sending the request!";
```
Comment (whimboo) [inline extension/chrome/content/nightly.js]: I would always start with something like: "Failed to submit data to %URL%: ". Then we can attach the reason as returned by the API.

**Comment by xabolcs** [comment]
@whimboo, regarding your comments about API: **there is no API**, IMHO.

**Comment by whimboo** [comment]
Well, we will get the response code with a description from the server for the POST request. In case of failures it should tell enough.

File: extension/chrome/content/nightly.js (hunk 4 above)
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: I disagree about leaving all up to `pastebin.mozilla.org`.

Comment (whimboo) [inline extension/chrome/content/nightly.js]: Then call at least "onError('No contents to post to pastebin.mozilla.org')" or similar.

Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Sure!

**Comment by xabolcs** [comment]
Sorry for the delay, I don't have always access to GitHub.
PR updated with commit a991c30d8596e3308496eef91de69b2d5972ee64.

File: extension/chrome/content/browser.js [hunk 8]
```diff
@@ -42,26 +42,35 @@ openURL: function(url)
   gBrowser.selectedTab = gBrowser.addTab(url);
 },
 
-openNotification: function(id, message, label, accessKey, callback) {
-  var action = {
-    label: label,
-    callback: callback,
-    accessKey: accessKey
+openNotification: function(id, message, label, accessKey, callback, boxPlease) {
```
Comment (xabolcs) [inline extension/chrome/content/browser.js]: I added the last parameter, `boxPlease` to let consumers choose the old, `notificationbox` style!
Suggestions about parameter name are welcome! :)

File: extension/chrome/content/browser.js [hunk 9]
```diff
@@ -42,26 +42,35 @@ openURL: function(url)
...
-      timeout: Date.now() + 10000
+      timeout: Date.now() + TIMEOUT,
+      removeOnDismissal: true
     };
 
-    PopupNotifications.show(gBrowser.selectedBrowser, id,
+    var notification = PopupNotifications.show(gBrowser.selectedBrowser, id,
       message, "urlbar", action, null, options);
+
+    setTimeout(function(){ notification.remove(); }, TIMEOUT);
```
Comment (xabolcs) [inline extension/chrome/content/browser.js]: Added automatic remove feature. :)

File: extension/chrome/content/browser.js [hunk 10]
```diff
@@ -42,26 +42,35 @@ openURL: function(url)
...
   } else {
     let nb = gBrowser.getNotificationBox();
 
-    nb.appendNotification(
+    var notification = nb.appendNotification(
       message, id,
       "chrome://nightly/content/brand/icon.png",
-      nb.PRIORITY_INFO_HIGH, [ action ]);
+      nb.PRIORITY_INFO_HIGH, action ? [ action ] : []);
+    setTimeout(function(){ nb.removeNotification(notification) }, TIMEOUT);
```
Comment (xabolcs) [inline extension/chrome/content/browser.js]: Same new feature here!

File: extension/chrome/content/nightly.js [hunk 11]
```diff
@@ -220,19 +220,55 @@ copyTemplate: function(template) {
   nightly.copyText(nightly.generateText(nightly.getTemplate(template)));
 },
 
-pastebin: function (content) {
-  var postdata = "paste_code=" + encodeURIComponent(content);
+pastebin: function (content, onLoadCallback, onErrorCallback) {
+  const PASTEBIN_URL = "https://pastebin.mozilla.orfg/";
```
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: This is a typo here, to test the `onError` stuff!
I'll fix this, once we deal with the `ProgressEvent`.

File: extension/chrome/content/nightly.js [hunk 12]
```diff
@@ -220,19 +220,55 @@ copyTemplate: function(template) {
...
     if (request.readyState == 4 ) {
-      if (request.status==200)
-        nightlyApp.openURL(request.responseText);
+      if (request.status === 200) {
+        nightlyApp.openURL(request.channel.URI.spec);
+      }
+      if (typeof onLoadCallback === "function") {
+        onLoadCallback();
+      }
+      console.log("pastebin done!");
```
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Just to debug.

File: extension/chrome/content/nightly.js [hunk 13]
```diff
@@ -220,19 +220,55 @@ copyTemplate: function(template) {
...
+
+  function onError(message) {
+    if (typeof onErrorCallback === "function") {
+      onErrorCallback();
+    }
+    if (message === null || message === "" || message === undefined) {
+      return;
+    }
+    if (typeof message !== "string") {
+      console.log("onerror! this is a ProgressEvent! Now what?", message);
```
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: @whimboo, please help me to extract nice details from a `ProgressEvent`!

File: extension/chrome/content/nightly.js [hunk 14]
```diff
@@ -220,19 +220,55 @@ copyTemplate: function(template) {
...
+      message = "Failed to submit data to " + PASTEBIN_URL + " !";
+    }
+    if (nightlyApp.openNotification) {
```
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: Messenger doesn't have implementation for it.

File: extension/chrome/content/nightly.js [hunk 15]
```diff
@@ -220,19 +220,55 @@ copyTemplate: function(template) {
...
+      nightlyApp.openNotification("nightly-pastebin-error", message, null, null, null, /* boxPlease */ true);
```
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: The old `notificationbox` bar is more noticable, IMHO

File: extension/chrome/content/browser.js (hunk 8 above)
Comment (whimboo) [inline extension/chrome/content/browser.js]: Can we do that in another PR. I don't see how this relates to the pastebin changes. Thanks.

File: extension/chrome/content/browser.js (hunk 9 above)
Comment (whimboo) [inline extension/chrome/content/browser.js]: Same here for a separate PR. Keep in mind that this would also end-up in the history then.

In general lets try to only do changes and additions to what's really needed for the appropriate issue. Adding more and more in following commits makes it hard to review.

File: extension/chrome/content/nightly.js (hunk 13 above)
Comment (whimboo) [inline extension/chrome/content/nightly.js]: Why do we have to listen for a progress event? Also the message parameter should really be a string only and nothing else. So no need to add all those checks. Please also move the call to onErrorCallback() after the console.log() call, so we keep the flow correct.

File: extension/chrome/content/browser.js (hunk 8 above)
Comment (xabolcs) [inline extension/chrome/content/browser.js]: Sure!
Reverting back.

File: extension/chrome/content/browser.js (hunk 9 above)
Comment (xabolcs) [inline extension/chrome/content/browser.js]: Sure!
Reverting back.

File: extension/chrome/content/nightly.js (hunk 13 above)
Comment (xabolcs) [inline extension/chrome/content/nightly.js]: > Why do we have to listen for a progress event?

Because of line [261 and 262][1]. Should I remove all `onError` & `onAbort` binding?


> Please also move the call to onErrorCallback() after the console.log() call, so we keep the flow correct.

I prefer calling callback first. Because the request already done. What if something goes wrong in ` nightlyApp.openNotification`?

[1]: https://github.com/mozilla/nightlytt/pull/93/files#diff-abc41711d41eef2b916d44abfc0d6af8R261

Comment (whimboo) [inline extension/chrome/content/nightly.js]: > Because of line 261 and 262. Should I remove all onError & onAbort binding?

I haven't used the XMLHttpRequest for a while, so I checked MDN and I cannot see those properties as part of the  class: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest

> I prefer calling callback first. Because the request already done. What if something goes wrong in nightlyApp.openNotification?

You allow a general callback, and you don't know what can happen inside it. openNotification just the case we use here for now. In case there is another callback which raises an exception or such, the console log entries are mixed up. I really want to avoid this.

**Comment by xabolcs** [comment]
This PR is outdated now. The new issue #251 need another PR.
